rasa/core/featurizers/tracker_featurizers.py

```diff
@@ -0,0 +1,356 @@
+        label_ids = self._convert_labels_to_ids(trackers_as_actions, domain)
+
+        return tracker_state_features, label_ids
+
+    def prediction_states(
+        self, trackers: List[DialogueStateTracker], domain: Domain
+    ) -> List[List[State]]:
+        """Transforms list of trackers to lists of states for prediction."""
+
+        raise NotImplementedError(
+            "Featurizer must have the capacity to create feature vector"
+        )
+
+    # noinspection PyPep8Naming
+    def create_state_features(
+        self,
+        trackers: List[DialogueStateTracker],
+        domain: Domain,
+        interpreter: NaturalLanguageInterpreter,
+    ) -> List[List[Dict[Text, List["Features"]]]]:
+        """Create X for prediction."""
+
+        trackers_as_states = self.prediction_states(trackers, domain)
+        return self._featurize_states(trackers_as_states, interpreter)
+
+    def persist(self, path) -> None:
+        featurizer_file = os.path.join(path, "featurizer.json")
+        rasa.utils.io.create_directory_for_file(featurizer_file)
+
+        # noinspection PyTypeChecker
+        rasa.utils.io.write_text_file(str(jsonpickle.encode(self)), featurizer_file)
+
+    @staticmethod
+    def load(path) -> Optional["TrackerFeaturizer"]:
+        """Loads the featurizer from file."""
+
+        featurizer_file = os.path.join(path, "featurizer.json")
+        if os.path.isfile(featurizer_file):
+            return jsonpickle.decode(rasa.utils.io.read_file(featurizer_file))
+
+        logger.error(
+            "Couldn't load featurizer for policy. "
+            "File '{}' doesn't exist.".format(featurizer_file)
+        )
+        return None
+
+
+class FullDialogueTrackerFeaturizer(TrackerFeaturizer):
+    """Creates full dialogue training data for time distributed architectures.
+
+    Creates training data that uses each time output for prediction.
+    Training data is padded up to the length of the longest dialogue with -1.
+    """
+
+    def training_states_and_actions(
+        self, trackers: List[DialogueStateTracker], domain: Domain
+    ) -> Tuple[List[List[State]], List[List[Text]]]:
+        """Transforms list of trackers to lists of states and actions.
+
+        Training data is padded up to the length of the longest dialogue with -1.
+        """
+
+        trackers_as_states = []
+        trackers_as_actions = []
+
+        logger.debug(
+            "Creating states and action examples from "
+            "collected trackers (by {}({}))..."
+            "".format(type(self).__name__, type(self.state_featurizer).__name__)
+        )
+        pbar = tqdm(trackers, desc="Processed trackers", disable=is_logging_disabled())
+        for tracker in pbar:
+            states = self._create_states(tracker, domain)
+
+            delete_first_state = False
+            actions = []
+            for event in tracker.applied_events():
+                if isinstance(event, ActionExecuted):
```

Comment:
let's use guards to lower the indentation level
```suggestion
                if not isinstance(event, ActionExecuted):
                    continue
```

Reply:
I don't agree here, I think direct `if` is more intuitive to grasp what's going on

Reply:
I agree with @wochinge 🙈

Reply:
We really need automated checks for that, then we can skip all the discussions, because then we can simply follow rules.

Reply:
following rules is not always the best idea, especially with research code

Reply:
but feel free to change it if you think it is better